src/libstd/path.rs

```diff
@@ -639,6 +639,18 @@ pub struct Iter<'a> {
     inner: Components<'a>,
 }
 
+#[stable(feature = "path_components_debug", since = "")]
+impl<'a> fmt::Debug for Components<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let components = self.as_path()
+                             .components()
+                             .collect::<Vec<_>>();
```

Comment:
It's a shame to have to allocate on debug. You should be able to do this using `debug_list` and a helper struct.